### PR TITLE
feat: abbreviate package.json semantic versions

### PR DESCRIPTION
**Pull request by tbjgolden**
#### Description

npm packages often use semantic release as a tool to automate versioning, by having a program look at conventional commit messages, like those on this repo.

This PR abbreviates long versions like `v0.0.0-semantically-released` to `semantic`

#### Motivation and Context

For npm packages, a somewhat common pattern is using semantic release, which allows a ci tool (e.g. github actions) to automatically determine versions based on (a conventional) commit history. This means that the correct version isn't actually in the package.json file.
Normally, you'll see a version like 0.0.0-semantically-released in the package.json, to indicate that the package version is derived elsewhere. In it's current implementation, this is awkward because it conveys little meaning, uses quite a bit of space, and doesn't actually let you know the version

#### Screenshots:

Screenshot_20210205_155513

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Review by vladimyr — APPROVED** [review]
This is nice!

> Normally, you'll see a version like 0.0.0-semantically-released in the package.json, to indicate that the package version is derived elsewhere. In it's current implementation, this is awkward because it conveys little meaning, uses quite a bit of space, and doesn't actually let you know the version

With that being said now you replaced the actual _version_ coming from `package.json` with a version hint but it lacks some visual indication to be easily differentiated from the regular version IMO

May I propose to emulate semantic release [badge](https://github.com/semantic-release/semantic-release#badge) by adding that rocket emoji before `semantic` 💡

File: src/modules/package.rs [hunk 1]
```diff
@@ -68,6 +68,10 @@ fn extract_package_version(file_contents: &str, display_private: bool) -> Option
     };
 
     let formatted_version = format_version(raw_version);
+    if formatted_version == "v0.0.0-development" || formatted_version.find("semantic").is_some() {
```
Comment (vladimyr) [inline src/modules/package.rs:71]: Let's make it bit more strict because we might encounter actual prerelase version (non-zero) containing `semantic` identifier:
```suggestion
    if formatted_version == "v0.0.0-development" || formatted_version.starts_with("v0.0.0-semantic") {
```

**Review by vladimyr — CHANGES_REQUESTED** [review]

File: src/modules/package.rs (hunk 1 above)
Comment (matchai) [reply src/modules/package.rs:71]: The current implementation is aiming to identify any of the following common semantic release placeholders:
- `0.0.0-semantically-released`
- `0.0.0-semantic-versioning`
- `0.0.0-semantic-release`

Do you know of any packages that currently have the `semantic` identifier in their version, and aren't using semantic-release?

Comment (vladimyr) [reply src/modules/package.rs:71]: > The current implementation is aiming to identify any of the following common semantic-release placeholders:

That's perfectly fine I just wanted to make it more strict to prevent something like `1.4.21-foo-semantic-bar` being incorrectly matched. And yes, I'm not aware of anyone out there using such ridiculous prerelease identifiers but let's not underestimate people's creativity :stuck_out_tongue_winking_eye:

Comment (tbjgolden) [reply src/modules/package.rs:71]: Was thinking about this when I wrote this originally actually

My thinking was if they had `semantic` in a version like this then it's arguably still more likely (albeit speculative) that  they were using semantic versioning than not with something like `1.4.21-foo-semantic-bar` - though not sure how we'd be able to explain the `1.4.21`

Open to tweak it or keep the same, what do you think @matchai ?

Comment (matchai) [reply src/modules/package.rs:71]: I think the safer option is to prefix with v0.0.0, and keep an eye out for issues opening up with exceptional edge cases 😊

File: src/modules/package.rs [hunk 2]
```diff
@@ -343,6 +347,51 @@ mod tests {
...
+        let config_name = "package.json";
+        let config_content = json::json!({
+            "name": "starship",
+            "version": "0.0.0-development"
+        })
+        .to_string();
+
+        let project_dir = create_project_dir()?;
+        fill_config(&project_dir, config_name, Some(&config_content))?;
+        expect_output(&project_dir, Some("semantic"), None)?;
```
Comment (vladimyr) [inline src/modules/package.rs:361]:
```suggestion
        expect_output(&project_dir, Some("semantic"), None);
```

File: src/modules/package.rs [hunk 3]
```diff
@@ -343,6 +347,51 @@ mod tests {
...
+        let config_name = "package.json";
+        let config_content = json::json!({
+            "name": "starship",
+            "version": "v0.0.0-semantically-released"
+        })
+        .to_string();
+
+        let project_dir = create_project_dir()?;
+        fill_config(&project_dir, config_name, Some(&config_content))?;
+        expect_output(&project_dir, Some("semantic"), None)?;
```
Comment (vladimyr) [inline src/modules/package.rs:376]:
```suggestion
        expect_output(&project_dir, Some("semantic"), None);
```

File: src/modules/package.rs [hunk 4]
```diff
@@ -343,6 +347,51 @@ mod tests {
...
+        let config_name = "package.json";
+        let config_content = json::json!({
+            "name": "starship",
+            "version": "v0.0.0-alpha"
+        })
+        .to_string();
+
+        let project_dir = create_project_dir()?;
+        fill_config(&project_dir, config_name, Some(&config_content))?;
+        expect_output(&project_dir, Some("v0.0.0-alpha"), None)?;
```
Comment (vladimyr) [inline src/modules/package.rs:391]:
```suggestion
        expect_output(&project_dir, Some("v0.0.0-alpha"), None);
```

**Review by vladimyr — CHANGES_REQUESTED** [review]

**Comment by matchai** [comment]
Hi @tbjgolden. It looks like this PR may have unfortunately fallen through the cracks. 😅 
I'd be happy to have it merged if you can go ahead and run `rustfmt` on the project.

**Comment by tbjgolden** [comment]
> Hi @tbjgolden. It looks like this PR may have unfortunately fallen through the cracks. 😅 
> I'd be happy to have it merged if you can go ahead and run `rustfmt` on the project

Ah yes - will fix up now

**Comment by tbjgolden** [comment]
Should be good to go

As for the rocket emoji to indicate semantic versioning - I like the idea though haven't included it here

Seems that the output currently generates a space between the package and the version, so might need some tweaking to get it showing `📦🚀` a la that badge

**Comment by vladimyr** [comment]
> Seems that the output currently generates a space between the package and the version, so might need some tweaking to get it showing `📦🚀` a la that badge

Using `\b` is certainly possible but hardly advisable. 🙃

**Review by vladimyr — APPROVED** [review]
LGTM 🚀

**Comment by matchai** [comment]
I think that the PR as-is serves the purpose of clarifying the version number well. I think that adding more emoji will start to clutter the prompt and confuse the user. I'm personally happy to merge it in its current state 😊

**Comment by vladimyr** [comment]
Sure, sure, that was just me rambling about possibilities. It is perfectly fine in its current form!

**Comment by matchai** [comment]
Thank you Tom and Dario for getting this PR though! 😊
